File: common_receiver.py
import numpy as np
def common_receiver(h,s,g):
    """
    h - offsets
    s - shots 
    g - geophone

    Given a common shot section, this code computes 
    all possible pair (s,h) such that g=s+2h, i.e., 
    a common receiver section

    Returned parameters are offsets and traces in-
    dexes to get the CR section from the CS data.
    """
    ds=s[1]-s[0]
    dh=h[1]-h[0]
    ratio=ds//dh
    shotind=(g-s[0])//ds
    gind=max(0,shotind - int((len(h)-1)/ratio)-h[0]//ds)
    geos=np.arange(gind,shotind-h[0]//ds)
    geos=geos*len(h) + ratio*np.arange(len(geos)-1,-1,-1)
    h=np.insert(h,0,-np.arange(len(geos)-1,-1,-1)*ratio*dh - h[0])

    
    return h,geos

File: commonreceiver.c
#include <stdio.h>
#include <stdlib.h>
#ifndef max
#define max(a,b) (((a)>(b))? (a) : (b))
#endif
/* Function CommonReceiver()
 *
 * 	Parameters:
 * 		s(input): 		Source coordinates 
 *		geos(output):	Index of shots for a common-receiver at position g
 *		h(input):		Offsets
 *		nh(input):		Number of different offsets
 *		g(input):		Coordinate of the common-receiver
 *
 *	Return:
 *		lengeos: Number of offsets in the common-receiver section
 *
 *	Description: Given sets of possible offsets and source positions in a seismic experiment 
 *				 this function computes possible source indexes, in such a way that we have a 
 *				 common-receiver section at g.
 * */


unsigned CommonReceiver(double* s,int* geos,double* h,double ds,double dh,unsigned nh,double g){
	double ratio;
	unsigned geoind,bigoffset,lengeos,smalloffset,ii;
	ratio=ds/dh;
	geoind=(g-s[0])/ds;
	smalloffset=geoind - (int)(h[0]/ds);
	bigoffset=max(0,smalloffset - (int)((nh-1)/ratio));
	lengeos=smalloffset - bigoffset+1;
	for(ii=0;ii<lengeos;ii++)
		geos[ii]=(bigoffset+ii)*nh + (lengeos-ii-1)*ratio;
	
	return lengeos;
}


